source/common/ssl/context_manager_impl.cc

```diff
@@ -20,23 +21,93 @@ void ContextManagerImpl::releaseContext(Context* context) {
+                                              const std::string& listener_name,
+                                              const std::vector<std::string>& server_names) {
+  std::unique_lock<std::mutex> lock(contexts_lock_);
+
+  // Remove mappings.
+  if (server_names.empty()) {
+    if (map_exact_[listener_name][EMPTY_STRING] == context) {
+      map_exact_[listener_name][EMPTY_STRING] = nullptr;
+    }
+  } else {
+    for (const auto& name : server_names) {
+      if (name.size() > 2 && name[0] == '*' && name[1] == '.') {
+        if (map_wildcard_[listener_name][name] == context) {
+          map_wildcard_[listener_name][name] = nullptr;
+        }
+      } else {
+        if (map_exact_[listener_name][name] == context) {
+          map_exact_[listener_name][name] = nullptr;
+        }
+      }
+    }
+  }
+
+  // context may not be found, in the case that a subclass of Context throws
+  // in it's constructor.  In that case the context did not get added, but
+  // the destructor of Context will run and call releaseContext().
+  contexts_.remove(context);
+}
+
 ClientContextPtr ContextManagerImpl::createSslClientContext(Stats::Scope& scope,
                                                             ClientContextConfig& config) {
-
   ClientContextPtr context(new ClientContextImpl(*this, scope, config));
   std::unique_lock<std::mutex> lock(contexts_lock_);
   contexts_.emplace_back(context.get());
   return context;
 }
 
-ServerContextPtr ContextManagerImpl::createSslServerContext(Stats::Scope& scope,
-                                                            ServerContextConfig& config) {
-  ServerContextPtr context(new ServerContextImpl(*this, scope, config, runtime_));
+ServerContextPtr
+ContextManagerImpl::createSslServerContext(const std::string& listener_name,
+                                           const std::vector<std::string>& server_names,
+                                           Stats::Scope& scope, ServerContextConfig& config) {
+  ServerContextPtr context(
+      new ServerContextImpl(*this, listener_name, server_names, scope, config, runtime_));
   std::unique_lock<std::mutex> lock(contexts_lock_);
   contexts_.emplace_back(context.get());
+
+  // Save mappings.
+  if (server_names.empty()) {
+    map_exact_[listener_name][EMPTY_STRING] = context.get();
+  } else {
+    for (const auto& name : server_names) {
+      if (name.size() > 2 && name[0] == '*' && name[1] == '.') {
+        map_wildcard_[listener_name][name] = context.get();
+      } else {
+        map_exact_[listener_name][name] = context.get();
+      }
+    }
+  }
+
   return context;
 }
 
+ServerContext* ContextManagerImpl::findSslServerContext(const std::string& listener_name,
+                                                        const std::string& server_name) {
+  std::unique_lock<std::mutex> lock(contexts_lock_);
+  if (map_exact_[listener_name][server_name] != nullptr) {
```

Comment:
In the case that server_name wasn't in the inner map, this inserts it, based on the user-provided SNI value in the handshake.  I think you need to use map_exact[listener_name].find(server_name) != end, and assume there are never nullptr's in the map.  Similar change needed in other places map_exact_ is used.

Reply:
Done.